crates/astria-composer/src/searcher/executor.rs

```diff
@@ -120,50 +124,59 @@ impl Executor {
+    async fn get_and_increment_nonce(&mut self) -> eyre::Result<u32> {
+        if let Some(curr_nonce) = self.nonce {
             self.nonce = Some(curr_nonce + 1);
-            curr_nonce
-        })
-    }
-
-    /// Sugmits a signed transaction to the sequencer node.
-    /// TODO: handle failed tx submission due to nonce
-    async fn submit_tx(&self, signed_tx: SignedTransaction) -> eyre::Result<()> {
-        let rsp = self
-            .sequencer_client
-            .submit_transaction_sync(signed_tx)
-            .await
-            .wrap_err("failed submitting transaction to sequencer")?;
-        if rsp.code.is_err() {
-            bail!(
-                "submitting transaction to sequencer returned with error code; code: `{code}`; \
-                 log: `{log}`; hash: `{hash}`",
-                code = rsp.code.value(),
-                log = rsp.log,
-                hash = rsp.hash,
-            );
+            Ok(curr_nonce)
+        } else {
+            let rsp = self.sequencer_client.get_latest_nonce(self.address).await?;
+            self.nonce = Some(rsp.nonce + 1);
+            Ok(rsp.nonce)
         }
-        Ok(())
     }
 
-    /// Signs and submits the bundle of actions to the sequencer.
-    async fn sign_and_submit(&mut self, actions: Vec<Action>) -> eyre::Result<()> {
+    /// Populates nonce, signs and submits the bundle of actions to the sequencer. Returns the
+    /// transaction if successfully submitted.
+    async fn execute_bundle(
+        &mut self,
+        bundle: Vec<Action>,
+    ) -> Result<SignedTransaction, ExeuctionError> {
         let nonce = self
             .get_and_increment_nonce()
-            .ok_or(eyre!("no nonce stored; cannot process bundle"))?;
+            .await
+            .map_err(ExeuctionError::SequencerClientFailure)?;
 
         let tx = UnsignedTransaction {
             nonce,
-            actions,
+            actions: bundle,
         }
         .into_signed(&self.sequencer_key);
 
-        self.submit_tx(tx)
+        let submission_rsp = self
+            .sequencer_client
+            .submit_transaction_sync(tx.clone())
             .await
-            .wrap_err("failed submitting signed actions to sequencer")?;
-        Ok(())
+            .map_err(ExeuctionError::SequencerClientFailure)?;
+
+        match AbciCode::from_tendermint(submission_rsp.code) {
+            Some(AbciCode::OK) => Ok(tx),
+            Some(AbciCode::INVALID_NONCE) => Err(ExeuctionError::InvalidNonce {
+                nonce,
+                transaction: tx,
+            }),
+            _ => Err(ExeuctionError::UnknownDeliverTxFailure {
+                code: submission_rsp.code,
+                transaction: tx,
+            }),
+        }
+    }
+
+    async fn handle_bundle_resubmission(&self, bundle: Vec<Action>) {
+        if let Err(e) = self.transaction_resubmission_tx.send(bundle).await {
+            error!(error.msg = %e, error.cause_chain=?e, "failed to resubmit transaction");
```

Comment:
```suggestion
            error!(error.msg = %e, error.cause_chain=?e, "transaction resubmission channel receiver dropped");
```

more accurate i think?